### ISSUE TITLE
File path too long?

### ISSUE THREAD

**Issue by harryray33** [dnGrep/dnGrep#309]
I've been searching a folder on my desktop with several subdirectories and keep getting error results along the lines of  'could not find a part of the path' marked in red.

I compressed the directory to a rar file and I managed to find the text I was looking for.

Is this possibly to do with file path or name being to long?

Below is example of one of the file paths

C:\Users\user\Desktop\31-10-18\01-11-18\Application confirmation_ *************** – *************** rooms with ****************** - **************.com - *****--01-11-2018\Application confirmation_ ******************************************************** - ****************** - *****--01-11-2018.txt
(The stars represent alpha numeric characters

**Comment by JVimes** [comment]
I think you're right. That path is 298 characters. Lots of Windows APIs can't handle more than 260 characters ([Maximum Path Length Limitation](https://docs.microsoft.com/en-us/windows/desktop/fileio/naming-a-file#maximum-path-length-limitation)). In the near term, I suggest dnGrep show a more descriptive error message when path length is the problem. And long-term, I hear there are ways around the MAX_PATH limit.

**Comment by harryray33** [comment]
Thanks, that does appear to be the problem. It's a complete pain!!

For individual directories I can compress  or move the directory from my desktop to the root drive but it's far from being a perfect solution and not always practical.

I'm going to run into problems if I need to do a disc wide search.

If anybody knows of a workaround I would appreciate the info.

**Comment by doug24** [comment]
What version of .net runtime do you have installed on that machine?  The [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.io.pathtoolongexception?view=netframework-4.7.2) says that .net 4.6.2 and later supports long paths.  But they are never clear if that means the target version at compile time, or the end user's runtime version.  dnGrep is currently uses .net 4.5.2 at compile time, and will run on any version 4.5.2 and higher.

The last time I worked on the directory and file enumeration code I specifically handled the PathTooLongException, so it would be interesting to see where it failed in the case.  I tried to test it and create a unit test for long paths, but it was just too difficult to create one - Windows Explorer is still limited to 260 characters on the latest Windows 10.

**Comment by harryray33** [comment]
Actually, I'm a bit confused about this...

According to the entry on services I have 4.0.30319 and according to my uninstall view I have the display name as 4.6.2 and a display name 4.6.01590 and I seem to have something slightly different in the registry.

Assuming I've looked at the correct key her'e the entry from the registry (sorry, for some reason I couldn't attach it as a file)

Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup]

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP]

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\CDF]

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\CDF\v4.0]
"WMIInstalled"=dword:00000001
"NetTcpPortSharingInstalled"=dword:00000001
"HttpNamespaceReservationInstalled"=dword:00000001
"NonHttpActivationInstalled"=dword:00000001
"SMSvcHostPath"="C:\\Windows\\Microsoft.NET\\Framework64\\v4.0.30319\\"

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727]
"Install"=dword:00000001
"Version"="2.0.50727.5420"
"Increment"="5420"
"SP"=dword:00000002
"CBS"=dword:00000001
"OCM"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1028]
"MSI"=dword:00000001
"Install"=dword:00000001
"OCM"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1029]
"Install"=dword:00000001
"OCM"=dword:00000001
"MSI"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1030]
"MSI"=dword:00000001
"Install"=dword:00000001
"OCM"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1031]
"OCM"=dword:00000001
"MSI"=dword:00000001
"Install"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1032]
"OCM"=dword:00000001
"Install"=dword:00000001
"MSI"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1033]
"Version"="2.0.50727.5420"
"CBS"=dword:00000001
"Increment"="5420"
"SP"=dword:00000002

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1035]
"Install"=dword:00000001
"OCM"=dword:00000001
"MSI"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1036]
"MSI"=dword:00000001
"Install"=dword:00000001
"OCM"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1038]
"OCM"=dword:00000001
"MSI"=dword:00000001
"Install"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1040]
"MSI"=dword:00000001
"Install"=dword:00000001
"OCM"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1041]
"MSI"=dword:00000001
"Install"=dword:00000001
"OCM"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1042]
"MSI"=dword:00000001
"Install"=dword:00000001
"OCM"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1043]
"OCM"=dword:00000001
"MSI"=dword:00000001
"Install"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1044]
"Install"=dword:00000001
"MSI"=dword:00000001
"OCM"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1045]
"OCM"=dword:00000001
"Install"=dword:00000001
"MSI"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1046]
"OCM"=dword:00000001
"Install"=dword:00000001
"MSI"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1049]
"Install"=dword:00000001
"MSI"=dword:00000001
"OCM"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1053]
"OCM"=dword:00000001
"Install"=dword:00000001
"MSI"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\1055]
"OCM"=dword:00000001
"MSI"=dword:00000001
"Install"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\2052]
"OCM"=dword:00000001
"Install"=dword:00000001
"MSI"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\2070]
"Install"=dword:00000001
"OCM"=dword:00000001
"MSI"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\3076]
"OCM"=dword:00000001
"MSI"=dword:00000001
"Install"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v2.0.50727\3082]
"OCM"=dword:00000001
"Install"=dword:00000001
"MSI"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v3.0]
"Version"="3.0.30729.5420"
"CBS"=dword:00000001
"Increment"="5420"
"Install"=dword:00000001
"SP"=dword:00000002

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v3.0\Servicing]

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v3.0\Servicing\Windows Workflow Foundation]
"SPIndex"=dword:00000000
"CBS"=dword:00000001
"Install"=dword:00000001
"SP"=dword:00000002
"SPName"="SP2"
"Hotfix"=""

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v3.0\Setup]
"Version"="3.0.30729.5420"
"InstallSuccess"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v3.0\Setup\1033]
"Version"="3.0.30729.5420"
"CBS"=dword:00000001
"Increment"="5420"
"InstallSuccess"=dword:00000001
"Install"=dword:00000001
"SP"=dword:00000002

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v3.0\Setup\Windows Communication Foundation]
"Version"="3.0.4506.5420"
"RuntimeInstallPath"="C:\\Windows\\Microsoft.NET\\Framework64\\v3.0\\Windows Communication Foundation\\"
"InstallSuccess"=dword:00000001
"ReferenceInstallPath"="C:\\Program Files\\Reference Assemblies\\Microsoft\\Framework\\v3.0\\"

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v3.0\Setup\Windows Presentation Foundation]
@="WPF v3.0.6920.5011"
"Version"="3.0.6920.5011"
"WPFCommonAssembliesPathx64"="C:\\Windows\\System32\\"
"InstallRoot"="C:\\Windows\\Microsoft.NET\\Framework64\\v3.0\\WPF\\"
"InstallSuccess"=dword:00000001
"WPFReferenceAssembliesPathx64"="C:\\Program Files\\Reference Assemblies\\Microsoft\\Framework\\v3.0\\"
"ProductVersion"="3.0.6920.5011"
"WPFNonReferenceAssembliesPathx64"="C:\\Windows\\Microsoft.NET\\Framework64\\v3.0\\WPF\\"

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v3.0\Setup\Windows Workflow Foundation]
@="Windows Workflow Foundation"
"InstallDir"="C:\\Program Files\\Reference Assemblies\\Microsoft\\Framework\\v3.0\\"
"MajorBuildNum"="4203"
"FileVersion"="3.0.4203.5420"
"InstallSuccess"=dword:00000001
"ProductVersion"="3.0.0.0"

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v3.5]
"Version"="3.5.30729.5420"
"CBS"=dword:00000001
"Install"=dword:00000001
"InstallPath"="C:\\Windows\\Microsoft.NET\\Framework64\\v3.5\\"
"SP"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v3.5\1033]
"Version"="3.5.30729.5420"
"CBS"=dword:00000001
"Install"=dword:00000001
"SP"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v4]

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Client]
"Version"="4.6.01590"
"TargetVersion"="4.0.0"
"Install"=dword:00000001
"MSI"=dword:00000001
"Servicing"=dword:00000000
"Release"=dword:00060636
"InstallPath"="C:\\Windows\\Microsoft.NET\\Framework64\\v4.0.30319\\"

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Client\1033]
"Version"="4.6.01590"
"TargetVersion"="4.0.0"
"Install"=dword:00000001
"Servicing"=dword:00000000
"Release"=dword:00060636

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Full]
"Version"="4.6.01590"
"TargetVersion"="4.0.0"
"Install"=dword:00000001
"MSI"=dword:00000001
"Servicing"=dword:00000000
"InstallPath"="C:\\Windows\\Microsoft.NET\\Framework64\\v4.0.30319\\"
"Release"=dword:00060636

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Full\1033]
"Version"="4.6.01590"
"TargetVersion"="4.0.0"
"Install"=dword:00000001
"Servicing"=dword:00000000
"Release"=dword:00060636

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v4.0]
@="deprecated"

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v4.0\Client]
"Version"="4.0.0.0"
"Install"=dword:00000001

**Comment by harryray33** [comment]
After a bit of research about entries in the registry re dot net online it appears that I have 4.62 installed. 
I noticed that there is a new 4.7.2 version.

Any idea if the 4.7.2 version will solve the problem? I'm loathe to install the new version because of the disc space it takes up but if it's a solution I'll install it.

I forgot to say that I'm running Win7.

May be worth mentioning that I have the Microsoft .Net framework NGEN v.4.0.3319_x86 and _x64 service disabled.
I don't know if this would cause a problem with the file name length.

Thanks.

**Comment by doug24** [comment]
Given that you already have 4.6.2 installed, I would doubt that installing a newer version would fix this issue.  If 4.6.2 really does support long paths, then dnGrep would need to be recompiled targeting at least 4.6.2.

**Comment by harryray33** [comment]
OK, thanks so much for your help.
Do you have any plans to do this, if not can you suggest any sort of workaround?

**Comment by doug24** [comment]
Actually, I just looked again, and the latest release of dnGrep is already targeting 4.6.2.  So the only thing I can do it is to try again to set up a test case and try running it in the debugger.  I may have time to look at it this weekend.

**Comment by harryray33** [comment]
That's great, thank you...

**Comment by doug24** [comment]
I did some more googling, and according to [this page](https://blogs.msdn.microsoft.com/jeremykuhne/2016/07/30/net-4-6-2-and-long-paths-on-windows-10/), long paths are only possible on Windows 10.  So even if I get it working in dnGrep, it still won't work on your Windows 7.   The .net code is calling the native Windows libraries, and it's not just a .net limitation.

So, do you have any control over the app that is writing those files?

**Comment by harryray33** [comment]
I can, do a degree, control where the files are written, for example write to C rather than desktop (which will help to cut the pathname a bit)but not always.
The problem is also that I have a lot of sub directories within the folders.

The problem is searching the many other files that have already been written.

If I'm searching just one directory I can copy the directory to my root drive which does help but not always or I can compress the folder with RAR and that seem to work but when searching many directories this isn't practical or possible.

I use DNGrep on a daily basis so it's obviously causing me a headache.

I'm not an expert (far from it) and I can't see any way round this. I suppose one way would be to upgrade to Win10 but that is something I don't want to do.

If there's anything you can do it would be greatly appreciated.

Thanks.

**Comment by JVimes** [comment]
Unfortunately, this is a tough request. Microsoft ended support for Windows 7 four years ago, and extended support ends 14 months from now. Any work we do for Windows 7 will risk bugs, help only a few users, and be obsolete soon. We'd have to be very careful, and the payoff is small. 🤔

**Comment by harryray33** [comment]
Thanks, I did see this about the path \\?\ but I wasn't sure how to implement it in DnGrep.

Anyway, I've now had time to try it and on the limited testing I've done on just one folder) it does appear to work). I also put a RAR compressed file in there as well.

There are a couple of oddities, it doesn't seem to find directory names, only file names and, for example if i search for 03-11-2018 it will find the file but if I enter 03-11 it wont find it, however if I tick the whole word box it finds both.
I've just noticed that the issue of finding 03-11 and directory doesn't work with the regular path either, do you know the reason?

You're right, the open containing folder and open doesn't work but if I open it in Emeditor (which I've set to default for .txt files) it works.

In general, my searches are just for text, either in text files or in .exe or .dll files (binary?) are the settings in the above screenshoot best?
I've found (I think) that just ticking the reges  options and not entering the path to match gives the same results (I think!)
Which setting would you recommend for text searching?

I hope all the above blurb makes sense to you :0)

**Comment by doug24** [comment]
> it doesn't seem to find directory names

Are you trying to use dnGrep to search for a directory?  As far as I know, there is no way to use dnGrep to search for directory names.  Other than searching for text in files, you can use dnGrep to search for files by leaving the 'Search for' box empty, and dnGrep will list all files matching the 'Paths to match' (there is an Option to enable/disable this feature).

Searching using plain text will be likely be faster than search using Regex.

To search .exe and .dll files, you need to have the 'Include binary files' checked.

If you really only want results from .txt, .exe and .dll files it will be faster to set the 'Paths to match' to `*.txt; *.exe; *.dll`

> search for 03-11-2018 it will find the file but if I enter 03-11 it wont find it

No idea what this problem is.  It works in a simple text file, it may depend on your specific file.

**Comment by harryray33** [comment]
No, not specifically, I use the programme Everything for searching for files and directories. I was just curious as to whether the directory names should show up.

Most of the time I don't search specific extensions so I will leave it as *.*. I just really needed to know which search was better and would pick up most matches.

I'm still trying to get a handle on Regex and I only really use it in Everything. I'm not really sure when I would have to use it in a Grep search.

I'm baffled about the 03-11 but I'll carry on experimenting.
The main thing is that there is a way round the file name length nonsense!!

Are you planning any further development on DnGrep? For me it seems to be the best Grep programme for what I need.
I did look at Powergrep but it's ridiculously expensive.

The only issue now is one that was discussed earlier regarding the Unicode and the autodetect setting but from what you said it looks like I will have to live with that one :0)

Thanks once again :0)

**Comment by doug24** [comment]
I didn't realize you were using `everything` for searches.  I haven't tested `everything` searches with long paths, but I guess you know the search syntax is different.

Yes, I'll probably do more development on dnGrep.  I use it everyday, and my main emphasis so far has been to fix things that I had problems with, or lots of users have had problems with.  I think most of those have been addressed now.  Since I'm doing this all on my own time, it has to be something I find interesting, and when I have some time to burn.  

I did start looking into third-party libraries to support long paths.  I've tried two so far - the first failed on the very first unit test I ran.  The second passed all the unit tests, but slowed down the searches, and that's not something I'm willing to do.  I'm working on a trying a third one now, but it is a lot of work to test.
